src/XamlX/IL/Emitters/PropertyAssignmentEmitter.cs

```diff
@@ -33,114 +32,279 @@ List<IXamlPropertySetter> ValidateAndGetSetters(XamlPropertyAssignmentNode an)
-                    }
+                for (var i = 0; i < an.Values.Count - 1; ++i)
+                    context.Emit(an.Values[i], codeGen, an.Values[i].Type.GetClrType());
+                context.Emit(dynamicValue, codeGen, dynamicValueType);
+                codeGen.EmitCall(method);
+            }
 
-                    IXamlLabel Next() => next ?? (next = codeGen.DefineLabel());
+            return XamlILNodeEmitResult.Void(1);
+        }
 
-                    var checkNext = false;
-                    if (setter.BinderParameters.AllowRuntimeNull)
-                        checkedTypes.Add(type);
-                    else
-                    {
-                        // Check for null; Also don't add this type to the list of checked ones because of the null check
-                        codeGen
-                            .Dup()
-                            .Brfalse(Next());
-                        checkNext = true;
-                    }
+        private static void RemoveRedundantSetters(IXamlType valueType, List<IXamlPropertySetter> setters)
+        {
+            if (setters.Count == 1)
+                return;
 
-                    // Only do dynamic checks if we know that type is not assignable by downcast 
-                    if (!type.IsAssignableFrom(value.Type.GetClrType()))
-                    {
-                        codeGen
-                            .Dup()
-                            .Isinst(type)
-                            .Brfalse(Next());
-                        checkNext = true;
-                    }
+            // If the value is a value type, always use the first one
+            if (valueType.IsValueType)
+            {
+                setters.RemoveRange(1, setters.Count - 1);
+                return;
+            }
 
-                    if (checkNext)
-                        hadJumps = true;
-                    
-                    ILEmitHelpers.EmitConvert(context, codeGen, value, value.Type.GetClrType(), type);
-                    context.Emit(setter, codeGen);
-                    if (hadJumps)
-                    {
-                        codeGen.Br(exit);
-                    }
+            for (int index = 0; index < setters.Count;)
+            {
+                var setter = setters[index];
+                var type = setter.Parameters.Last();
 
-                    if(!checkNext)
-                        break;
+                // the value is directly assignable by downcast and the setter allows null: it will always match
+                if (type.IsAssignableFrom(valueType) && setter.BinderParameters.AllowRuntimeNull)
+                {
+                    setters.RemoveRange(index + 1, setters.Count - index - 1);
+                    return;
+                }
+
+                // the value type has already been handled by a previous setter
+                if (setters.Take(index).Any(previous => IsAssignableToWithNullability(setter, previous)))
```

Comment:
The comment is a bit confusing, you probably mean "the setter that's directly compatible with the type of the value is handled by the previous check"

Reply:
I agree it's confusing. However it doesn't reference the previous check directly (the static downcast one) but any previous runtime check. 

There was already one comment previously that was lost during the refactor: *We have already checked this type or its base type* which is clearer, I'll put it back instead if that's ok with you.